src/app/products/edit-product.component.ts

```diff
@@ -0,0 +1,153 @@
+    },
+    country: {
+      required: 'Country is required.'
+    }
+  };
+
+  // Define an instance of the validator for use with this form,
+  // passing in this form's set of validation messages.
+  this.genericValidator = new GenericValidator(this.validationMessages);
+}
+
+  ngOnInit(): void {
+    this.productForm = this.fb.group({
+      productName: ['', [Validators.required,
+                         Validators.minLength(3),
+                         Validators.maxLength(50)]],
+      productCode: ['', Validators.required],
+      starRating: ['', NumberValidators.range(1, 5)],
+      tags: this.fb.array([]),
+      description: '',
+      productType :['Type1',Validators.required],
+      country:['',Validators.required],
+      price:['',[Validators.required,NumberValidators.range(1,1000)]],
+    });
+
+    this.sub = this.route.paramMap.subscribe(
+      params => {
+       this.id = +Number(params.get('id'));
+        this.getProduct();
+      }
+    );
+    this.sub1 =   this.productService.getCountries().subscribe({
+      next: countries => {
+        this.countries  = countries;
+      },
+      error: err => this.errorMessage = err
+    });
+  }
+
+  ngOnDestroy(): void {
+    this.sub.unsubscribe();
+    this.sub1.unsubscribe();
+  }
+
+  ngAfterViewInit(): void {
+    // Watch for the blur event from any input element on the form.
+    // This is required because the valueChanges does not provide notification on blur
+    const controlBlurs: Observable<any>[] = this.formInputElements
+      .map((formControl: ElementRef) => fromEvent(formControl.nativeElement, 'blur'));
+
+    // Merge the blur event observable with the valueChanges observable
+    // so we only need to subscribe once.
+    merge(this.productForm.valueChanges, ...controlBlurs).pipe(
+      debounceTime(800)
+    ).subscribe(value => {
+      this.displayMessage = this.genericValidator.processMessages(this.productForm);
+    });
+  }
+
+  getProduct(): void {
+    if(this.id >0)
+    {
+    let products : IProduct[] =JSON.parse(localStorage.getItem("my-products") || '{}');
+    if(products!=null && products.length!=0 && products.length!=undefined)
+    {
+      this.product = products.find(t=>t.id==this.id)!;
+      // Update the data on the form
+    this.productForm.patchValue({
+      productName: this.product.productName,
+      productCode: this.product.productCode,
+      starRating: this.product.starRating,
+      description: this.product.description,
+      country :this.product.country,
+      price : this.product.price,
+      productType: (this.product.productType == null || this.product.productType == '' || this.product.productType == undefined) ? 'Type1' : this.product.productType
+    });
+    }
+    this.pageTitle = `Edit Product: ${this.product.productName}`;
+  }
+  }
+
+  saveProduct() : void{
+    if (this.productForm.valid) {
+      if (this.productForm.dirty) {
+        const p = { ...this.product, ...this.productForm.value };
```

Comment:
what is ...

Reply:
That is the code to copy productform value to product, this code i got from some tutorials